src/main/java/io/vertx/core/http/impl/SharedClientHttpStreamEndpoint.java

```diff
@@ -0,0 +1,157 @@
+import io.vertx.core.net.impl.pool.ConnectionPool;
+import io.vertx.core.net.impl.pool.PoolConnector;
+import io.vertx.core.net.impl.clientconnection.Lease;
+import io.vertx.core.net.impl.pool.PoolWaiter;
+import io.vertx.core.spi.metrics.ClientMetrics;
+
+import java.util.List;
+
+/**
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+class SharedClientHttpStreamEndpoint extends ClientHttpEndpointBase<Lease<HttpClientConnection>> implements PoolConnector<HttpClientConnection> {
+
+  private final HttpClientImpl client;
+  private final HttpChannelConnector connector;
+  private final ConnectionPool<HttpClientConnection> pool;
+  private final int http1MaxSize;
+  private final int http2MaxSize;
+
+  public SharedClientHttpStreamEndpoint(HttpClientImpl client,
+                                        ClientMetrics metrics,
+                                        int queueMaxSize,
+                                        int http1MaxSize,
+                                        int http2MaxSize,
+                                        HttpChannelConnector connector,
+                                        Runnable dispose) {
+    super(metrics, dispose);
+    this.client = client;
+    this.connector = connector;
+    this.http1MaxSize = http1MaxSize;
+    this.http2MaxSize = http2MaxSize;
+    this.pool = ConnectionPool.pool(this, Math.max(http1MaxSize, http2MaxSize), http1MaxSize * http2MaxSize, queueMaxSize);
+  }
+
+  @Override
+  public void connect(EventLoopContext context, Listener listener, Handler<AsyncResult<ConnectResult<HttpClientConnection>>> handler) {
+    connector
+      .httpConnect(context)
+      .onComplete(ar -> {
+        if (ar.succeeded()) {
+          incRefCount();
+          HttpClientConnection connection = ar.result();
+          connection.evictionHandler(v -> {
+            decRefCount();
+            listener.onRemove();
+          });
+          connection.concurrencyChangeHandler(listener::onCapacityChange);
+          long capacity = connection.concurrency();
+          Handler<HttpConnection> connectionHandler = client.connectionHandler();
+          if (connectionHandler != null) {
+            context.emit(connection, connectionHandler);
+          }
+          int weight;
+          if (connection instanceof Http1xClientConnection) {
+            weight = http2MaxSize;
+          } else {
+            weight = http1MaxSize;
+          }
+          handler.handle(Future.succeededFuture(new ConnectResult<>(connection, capacity, weight)));
+        } else {
+          handler.handle(Future.failedFuture(ar.cause()));
+        }
+      });
+  }
+
+  @Override
+  public boolean isValid(HttpClientConnection connection) {
+    return connection.isValid();
+  }
+
+  void checkExpired() {
+    pool.evict(conn -> !conn.isValid(), ar -> {
+      if (ar.succeeded()) {
+        List<HttpClientConnection> lst = ar.result();
+        lst.forEach(HttpConnection::close);
```

Comment:
Since `pool.evict()` removes the connections from pool, and  `HttpConnection.close()` tries to remove it from pool again via the `evictHandler, (listener.onRemove())`, does it make sense to mark the evicted connections as `evicted = true` here, and call `decRefCount()` to keep the ref counts consistent in the Endpoint ?

Reply:
I don't think so because Http1xClientConnection/Http2ClientConnection have an `evicted` flag to do this.

Reply:
I think we could add a test for HTTP/1.x that uses a short keep alive timeout to evict connection from the pool and when connection is closed, a new request is made to create a new connection. This way we can ensure that the endpoint is never depleted ?

Reply:
> I don't think so because Http1xClientConnection/Http2ClientConnection have an evicted flag to do this.

When the connection is evicted using `pool.evict()`, the `evicted` flag of the connection is still `false`, because the `evictHandler` is not invoked yet, it is called on `HttpConnection::close`

Reply:
> I think we could add a test for HTTP/1.x that uses a short keep alive timeout to evict connection from the pool and when connection is closed, a new request is made to create a new connection. This way we can ensure that the endpoint is never depleted ?

Sorry that I may not make it clear, I think there is a [test case](https://github.com/eclipse-vertx/vert.x/pull/3731/commits/eb2f9427a2c2e4e316dc6d7e6331fa959d44af3a) to cover what I concerned.

It does not cause issues in mail client, I think it is because the [listernr.onRemove()](https://github.com/eclipse-vertx/vert.x/pull/3731/files#diff-7695048cda6da12d4c94190f135d7f25e864955644b466403e5179849f03d00cR65) doesn't do anything since[ pool.slots[removed.index] != removed](https://github.com/eclipse-vertx/vert.x/pull/3731/files#diff-47e2e3f2721793f39461349a5aa942a0f36118b3bcd405044cc2e62e9afbdc56R290) is true.

So we are good :)

Reply:
Yes this case can happen but it is expected. Thanks for the feedback.

BTW if you can review the internal pool actions to find bugs, it would help a lot :-)